kvserver: propagate GCHint for partial deletions

pav-kv opened the pull request:
Previously, `GCHint` was used for 3 purposes:
1. Instruct GC to run at specific times. This is used by SQL table/index drop jobs, to limit the amount of time they wait on data GC.
2. Hint GC that a range is fully covered by range tombstones, so that GC can use `ClearRange` for faster deletes. This is useful for bulk data deletions as well.
3. Hint GC that a range was fully covered by range tombstones at some recent point. GC prioritizes going through such ranges, and processes them together, because this makes Pebble compaction more efficient.

The use-case 1 was broken. If a range tombstone does not fully cover the `Range` keyspace, `GCHint` is not written. Correspondingly, for small table/index deletions (spanning less than a single range), or deletions that don't perfectly match range bounds, there will be some ranges with no hints. For these ranges, GC might be arbitrarily delayed, and the schema change jobs would be stuck.

In addition, `GCHint` propagation during merges was lossy. It could drop the hint if either LHS or RHS doesn't have a hint, or has some data.

This commit extends `GCHint` to fix the use-case 1. Two fields are added: `min` and `max` deletion timestamp. These fields are populated even for partial range deletions, and are propagated during splits and merges unconditionally. The choice of min-max approach is explained in the comments to the corresponding `GCHint` fields.

Fixes #107931

Release note (bug fix): Fixed a bug that could occasionally cause schema change jobs (e.g. table/index drops) to appear stuck in state "waiting for MVCC GC" for much longer than expected. The fix only applies to future schema changes -- existing stuck jobs can be processed by manually force-enqueueing the relevant ranges in the MVCC GC queue under the DB Console's advanced debug page.

Epic: none

pav-kv committed:
kvserver: propagate GCHint for partial deletions

Previously, GCHint was used for 3 purposes:
1. Instruct GC to run at specific times. This is used by SQL table/index
   drop jobs, to limit the amount of time they wait on data GC.
2. Hint GC that a range is fully covered by range tombstones, so that GC
   can use ClearRange for faster deletes. This is useful for bulk data
   deletions as well.
3. Hint GC that a range was fully covered by range tombstones at some
   recent point. GC prioritizes going through such ranges, and processes
   them together, because this makes Pebble compaction more efficient.

The use-case 1 was broken. If a range tombstone does not fully cover the
Range keyspace, GCHint is not written. Correspondingly, for small
table/index deletions (spanning less than a single range), or deletions
that don't perfectly match range bounds, there will be some ranges with
no hints. For these ranges, GC might be arbitrarily delayed, and the
schema change jobs would be stuck.

In addition, GCHint propagation during merges was lossy. It could drop
the hint if either LHS or RHS doesn't have a hint, or has some data.

This commit extends GCHint to fix the use-case 1. Two fields are added:
min and max deletion timestamp. These fields are populated even for
partial range deletions, and are propagated during splits and merges
unconditionally. The choice of min-max approach is explained in the
comments to the corresponding GCHint fields.

Release note (bug fix): Fixed a bug that could occasionally cause schema
change jobs (e.g. table/index drops) to appear stuck in state "waiting
for MVCC GC" for much longer than expected. The fix only applies to
future schema changes -- existing stuck jobs can be processed by
manually force-enqueueing the relevant ranges in the MVCC GC queue under
the DB Console's advanced debug page.

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110078)
<!-- Reviewable:end -->

pavelkalinnikov commented:
`TestStatsAreDeletedForDroppedTables` panics in raft apply stack. It sees inconsistent GC threshold on disk and in memory.

```
F230906 18:21:09.717973 99124 kv/kvserver/replica.go:1600  [T1,Vsystem,n1,s1,r65/1:/Table/10{4-5},raft] 1  on-disk and in-memory state diverged: [GCThreshold.WallTime: 1694024462607440629 != 0]
```

pav-kv committed:
kvserver: document the usage of GCHint

The GCHint is currently used for 3 purposes. Document the status quo
before making changes.

Epic: none
Release note: none

pav-kv committed:
kvserver: add GCHint.UpdateAfterGC method

This is a cosmetic commit. It replaces a lower-level GCHint "reset"
method with a method that takes an effective GC threshold.

Epic: none
Release note: none

pavelkalinnikov commented on pkg/kv/kvserver/batcheval/cmd_delete_range.go:
```diff
@@ -123,21 +123,24 @@ func DeleteRange(
-			if !hint.ForwardLatestRangeDeleteTimestamp(h.Timestamp) {
+
+			// Add the timestamp to GCHint to guarantee that GC eventually clears it.
+			updated := hint.AddDeleteTimestamp(h.Timestamp)
```
I think we could use this hint independently of range tombstones and range deletions. Not doing for now, to keep the changes minimal.

pavelkalinnikov commented on pkg/kv/kvserver/mvcc_gc_queue.go:
```diff
@@ -501,14 +501,16 @@ func makeMVCCGCQueueScoreImpl(
 	}
 
+	if shouldRunGC(hint, gcTTL, now) {
+		r.ShouldQueue = true
```
I don't have a good feel whether this should instead be:
```golang
r.ShouldQueue = canAdvanceGCThreshold
```

Some code above sets `ShouldQueue` unconditionally, even if `canAdvanceGCThreshold` is false, some doesn't.

erikgrinaker replied:
This should be canAdvanceGCThreshold, I think. The other callers above that set it unconditionally do it to clean up intents and abort spans for abandoned transactions, which can always be removed regardless of the GC threshold.

pavelkalinnikov replied:
Done.

pavelkalinnikov commented:
@erikgrinaker PTAL. First 2 commits are comments/no-op. The real change is in the last commit.

Could add/extend some tests too. For now I've only manually verified that this fixes the repro in #107931.

pavelkalinnikov commented:
@erikgrinaker For the last commit, I suggest reviewing the `proto` file first, to get a feel for the algorithm. Then the `metadata.go` file to see how the `GCHint` is updated. Then the rest of it: the flow starts in `cmd_delete_range.go` (the hint is written or updated), goes through `cmd_end_transaction.go` (splits/merges), then via `mvcc_gc_queue.go` (where the GC decision is made), and `cmd_gc.go` where we finally GC and update the hint agian.

pavelkalinnikov commented on pkg/roachpb/metadata.go:
```diff
@@ -916,7 +916,7 @@ func (l Locality) AddTier(tier Tier) Locality {
+	return h.LatestRangeDeleteTimestamp.IsEmpty() && h.MinDeleteTimestamp.IsEmpty()
 }
 
 // Merge combines GC hints of two ranges. The result is either a hint that
```
Need to update this comment too.

pavelkalinnikov replied:
Done.

pavelkalinnikov commented on pkg/roachpb/metadata.go:
```diff
@@ -925,23 +925,26 @@ func (h *GCHint) IsEmpty() bool {
-		updated := h.LatestRangeDeleteTimestamp.IsSet()
+		updated = updated || h.LatestRangeDeleteTimestamp.IsSet()
 		h.LatestRangeDeleteTimestamp = hlc.Timestamp{}
-		return updated
```
Should still `return updated` here, we don't want to call `ForwardLatestRangeDeleteTimestamp` in this case.

Need to better test this `Merge` method.

pavelkalinnikov replied:
Done.

erikgrinaker commented on pkg/roachpb/metadata.go:
```diff
@@ -946,7 +957,50 @@ func (h *GCHint) ForwardLatestRangeDeleteTimestamp(ts hlc.Timestamp) bool {
-	h.LatestRangeDeleteTimestamp = hlc.Timestamp{}
+// AddDeleteTimestamp adds the given deletion timestamp into the hint. Returns
+// true iff the hint was updated.
+func (h *GCHint) AddDeleteTimestamp(ts hlc.Timestamp) bool {
```
Maybe something like `ScheduleGCAt` instead, to make it clear that the caller wants to run GC at some time, and the hint will figure out the best way to handle that.

`AddDeleteTimestamp` implies that the caller would only ever want to do this when it deletes something, but that isn't true.

pavelkalinnikov replied:
Done.

erikgrinaker commented on pkg/roachpb/metadata.go:
```diff
@@ -946,7 +957,50 @@ func (h *GCHint) ForwardLatestRangeDeleteTimestamp(ts hlc.Timestamp) bool {
+	if h.MinDeleteTimestamp.IsEmpty() {
+		h.MinDeleteTimestamp = ts
+	} else if cmp := h.MinDeleteTimestamp.Compare(ts); cmp > 0 {
+		h.MinDeleteTimestamp = ts
```
I think this needs to move the old MinDeleteTimestamp to MaxTimestamp if it's empty, otherwise we won't GC it eagerly.

pavelkalinnikov replied:
Good catch. Done this + exhaustive test that would have caught it.

erikgrinaker commented on pkg/kv/kvserver/mvcc_gc_queue.go:
```diff
@@ -500,14 +501,16 @@ func makeMVCCGCQueueScoreImpl(
+	}
+
 	maybeRangeDel := suspectedFullRangeDeletion(ms)
 	hasActiveGCHint := gcHintedRangeDelete(hint, gcTTL, now)
```
Let's rename this too. It may have a hint even if it doesn't have a range delete hint.

We may also want to keep the `r.ShouldQueue` here, and handle the range-delete hint backwards compatibility here until 24.1.

pavelkalinnikov replied:
Done.

erikgrinaker commented on pkg/kv/kvserver/mvcc_gc_queue.go:
```diff
@@ -520,12 +523,26 @@ func makeMVCCGCQueueScoreImpl(
+}
+
+// shouldRunGC returns whether the GCHint suggests running GC.
+func shouldRunGC(hint roachpb.GCHint, ttl time.Duration, now hlc.Timestamp) bool {
```
This name is too general, since there's many other conditions that also come into whether twe should run GC. In fact, since we only really need to check `MinDeleteTimestamp` if we keep the old range delete check as-is, then we can just inline `if canGC(MinDeleteTimestamp) { ShouldQueue: true }` and drop this function.

pavelkalinnikov replied:
Done.

pav-kv committed:
roachpb: add unit tests for GCHint.Merge

The test checks that all the GCHint invariants hold before/after the
Merge operation; that the Merge operation is commutative; and that Merge
returns true iff it modified the hint.

For now, the test only Merges an empty hint with a non-empty hint. Soon
adding tests for non-empty + non-empty cases.

The test helped to fix one bug in Merge commutativity.

Epic: none
Release note: none

erikgrinaker reviewed (commented):
First pass looks good, mostly nits. Thanks for picking this up!

I need to do another pass and just make sure the merge/split handling and such is good, will revisit. Need unit tests for some of this logic for sure.

What's your sense on backwards compatibility and backporting here? Do we want to make sure this works in mixed-version states and backport it to 23.1? Or are there any other less-risky mitigations we can do for 23.1 that doesn't involve protocol changes, and we'll version gate this change?

pavelkalinnikov commented on pkg/roachpb/metadata.proto:
```diff
@@ -463,7 +463,87 @@ message Version {
+  //
+  // When this hint is set, it is likely that adjacent ranges have it too,
+  // especially in bulk deletion situations. Once the GC queue sees this hint,
+  // it scans meta to find all the ranges that have it set too, and GC them with
```
Done.

pav-kv reviewed (commented):
Addressed some comments, still working on others.

erikgrinaker commented on pkg/roachpb/metadata.proto:
```diff
@@ -463,7 +463,87 @@ message Version {
+  //
+  // Initially, when a large key span is deleted, this hint is accurate for all
+  // the affected ranges. As the time goes, more data can be written to these
+  // ranges, or they may be split/merged with other ranges. We do best effort to
```
The splits don't matter though, only whether the first/last ranges merge into their neighbours, regardless of what happens in the meanwhile. Nbd though, don't feel strongly here.

erikgrinaker commented on pkg/kv/kvpb/api.proto:
```diff
@@ -457,9 +457,10 @@ message DeleteRangeRequest {
-  // If enabled and DeleteRangeRequest is deleting the whole range with a range
-  // tombstone, GCHint on replica will be updated to notify mvcc gc queue
-  // that this range would benefit from optimized GC process.
+  // If enabled, and a span is deleted using a range tombstone, then the GCHint
+  // on the corresponding Range will be updated. The hint instructs MVCC GC
+  // queue to delete this data as soon as it can, and helps optimizing GC for
+  // bulk deletions.
```
Hm, not sure what I was thinking here. Maybe I was looking at an intermediate commit or something.

pavelkalinnikov commented on pkg/roachpb/metadata.proto:
```diff
@@ -463,7 +463,87 @@ message Version {
+  // timestamp. The only remaining timestamp (max) is min+TTL. The next possible
+  // expedited GC run is thus at min+2*TTL. We may have omitted some timestamps
+  // between min and max, and they are now <= 2*TTL away from the next GC run.
+  optional util.hlc.Timestamp min_delete_timestamp = 2 [(gogoproto.nullable) = false];
```
How do we better distinguish between the timestamp of the data and the timestamp of the GC run? `gc_at` sounds like "when to run GC", which a) is >= TTL away from the the data deletion timestamp, and b) is not "stable", as the TTL may change.

So I'd like the name to convey the "data timestamp" semantics. `GCFor[Next]` / `GCUpTo[Next]` / `GCTimestamp[Next]`? I'm leaning towards the last option.

pavelkalinnikov replied:
Renamed to `GCTimestamp` and `GCTimestampNext`.

pav-kv committed:
roachpb: unit tests for GCHint.ScheduleGCFor

The test checks that all the GCHint invariants hold before/after the
ScheduleGCFor operation, and that it returns true iff it modified the
hint.

The test helped to fix one bug in the method implementation.

Epic: none
Release note: none

pav-kv committed:
roachpb: unit test for GCHint.UpdateAfterGC

Epic: none
Release note: none

pav-kv committed:
roachpb: test GCHint.ForwardLatestRangeDeleteTimestamp

Epic: none
Release note: none

pav-kv committed:
roachpb: more tests for GCHint.Merge

Epic: none
Release note: none

pavelkalinnikov commented:
@erikgrinaker

> I need to do another pass and just make sure the merge/split handling and such is good, will revisit. Need unit tests for some of this logic for sure.

Added unit tests and addressed most comments (still working on a couple minor ones). Ready for another review.

> What's your sense on backwards compatibility and backporting here? Do we want to make sure this works in mixed-version states and backport it to 23.1? Or are there any other less-risky mitigations we can do for 23.1 that doesn't involve protocol changes, and we'll version gate this change?

The new code handles old hints trivially: it respects the "whole range deleted" hint the same way as the old code, and does nothing for the "GC pending" hint because it's missing. The old code handles new hints by respecting the "whole range deleted" part of it, and simply ignoring the other fields.

We can only say that the GC is guaranteed for deletions happened in a cluster under the new version. With mixed-version clusters, a hint may be rewritten and lose "precision".

I think we don't need a version gate, and can backport to 23.1.

Alternative approaches:
- Always run GC. E.g. is last GC run was K * TTL ago, run it unconditionally. Obv., this has cost, especially on clusters with large cold datasets.
- Make the schema change job (that waits on GC) actively tell GC which ranges to process rather than do it once, and then sit and hope that it happens. This needs some plumbing. Doesn't seem like a better approach.

erikgrinaker commented on pkg/roachpb/metadata.proto:
```diff
@@ -463,7 +463,94 @@ message Version {
+  //
+  // NB: if GCTimestamp is set and GCTimestampNext is empty, it means there is
+  // only one pending timestamp for garbage collection.
+  optional util.hlc.Timestamp next_gc_timestamp = 3 [(gogoproto.nullable) = false,
```
nit: the field name (`next_gc_timestamp`) does not match the custom name and description (`GCTimestampNext`). Let's unify them -- don't particularly care which one we choose.

pavelkalinnikov replied:
Good catch. Renamed to `gc_timestamp_next`, as I initially intended.

erikgrinaker commented on pkg/roachpb/metadata.go:
```diff
@@ -946,7 +956,53 @@ func (h *GCHint) ForwardLatestRangeDeleteTimestamp(ts hlc.Timestamp) bool {
-// ResetLatestRangeDeleteTimestamp resets delete range timestamp.
-func (h *GCHint) ResetLatestRangeDeleteTimestamp() {
-	h.LatestRangeDeleteTimestamp = hlc.Timestamp{}
+// ScheduleGCFor updates the hint to schedule GC for the given timestamp.
```
```suggestion
// ScheduleGCFor updates the hint to schedule eager GC for data up to the given
// timestamp. When this timestamp falls below the GC threshold/TTL, it will
// be eagerly enqueued for GC when considered by the MVCC GC queue.
```

erikgrinaker commented on pkg/roachpb/metadata.go:
```diff
@@ -946,7 +956,53 @@ func (h *GCHint) ForwardLatestRangeDeleteTimestamp(ts hlc.Timestamp) bool {
-	h.LatestRangeDeleteTimestamp = hlc.Timestamp{}
+// ScheduleGCFor updates the hint to schedule GC for the given timestamp.
+// Returns true iff the hint was updated.
+func (h *GCHint) ScheduleGCFor(ts hlc.Timestamp) bool {
```
Maybe this is excessive, but is it worth passing in `now` here and clamping the timestamp to it, to guard against future timestamps? If this happened, it would ruin the 2xTTL worst-case guarantee. Then again, we can probably assume these timestamps to have interacted with the HLC, and if we ever got future timestamps into the HLC we're pretty screwed anyway. Probably safe to drop, just leaving it for your consideration.

erikgrinaker replied:
Meh, let's drop this -- if we have far-future HLC timestamps then we have much bigger problems than slow GC.

erikgrinaker commented on pkg/kv/kvserver/mvcc_gc_queue.go:
```diff
@@ -500,10 +501,20 @@ func makeMVCCGCQueueScoreImpl(
+		// now, however we should probably check a stronger condition that the GC
+		// threshold can be advanced to timestamp >= hint.GCTimestamp. The current
+		// way can result in false positives and wasteful GC runs.
+		r.ShouldQueue = canAdvanceGCThreshold
```
Is it worth increasing the score here to give this range priority over other queued ranges, in the case of a backlog? Maybe not, since other ranges with higher scores have actual garbage we want to get rid of, but if we do we could consider bumping this by 2 or something (i.e. 2x `mvccGCKeyScoreThreshold`). I think it's probably ok to leave this as-is though.

pavelkalinnikov replied:
Leaving as is, because the testing went good. Can revisit the priorities later.

erikgrinaker commented on pkg/kv/kvserver/client_merge_test.go:
```diff
@@ -5266,14 +5266,14 @@ func TestStoreMergeGCHint(t *testing.T) {
 			dataRight:  true,
 			delLeft:    true,
-			expectHint: false,
+			expectHint: true,
```
We should probably extend this test to assert what it originally wanted to assert, i.e. the `LatestRangeDeleteTimestamp` hint. We can add an additional assertion that it propagates the `GCTimestamp` hint.

pavelkalinnikov replied:
Done.

erikgrinaker reviewed (commented):
Nice work! I'm going to do a pass over the merge/split and below-Raft handling of the hint, to make sure this is safe in mixed versions, but I think this is probably good to go.

Thanks again for picking this up.

>Release note (bug fix): fixed a bug in GCHint causing unbounded wait for GC in schema change jobs, e.g. on table/index drop operations. After the change, the GCHint is populated and propagated properly, and eventual / timely (order of TTL) garbage collection is guaranteed.

Users don't know what GCHint is or why they should care. Try something like:

Release note (bug fix): Fixed a bug that could occasionally cause schema change jobs (e.g. table/index drops) to appear stuck in state "waiting for MVCC GC" for much longer than expected. The fix only applies to future schema changes -- existing stuck jobs can be processed by manually force-enqueueing the relevant ranges in the MVCC GC queue under the DB Console's advanced debug page.

erikgrinaker reviewed (approved):
Afaict, this should be good to go.

Another case we should test is sending a snapshot from a new node to an old node, and vice versa. Can be done with `ALTER RANGE ... RELOCATE`. Also, running consistency checks on the range.

pavelkalinnikov commented:
Verified that after all the changes this PR still fixes the repro test from #107931.

pav-kv committed:
kvserver: cover more cases in TestStoreMergeGCHint

Epic: none
Release note: none

pavelkalinnikov commented:
> What's your sense on backwards compatibility and backporting here? Do we want to make sure this works in mixed-version states and backport it to 23.1?

I did a couple of mixed-version tests manually, with before/after this PR (let's call them `vBefore` and `vAfter`). Works as expected, and even better.

For example, one of the tests is a modification of #107931 repro:

1. Start a `vAfter` cluster, create a table+index, write some values.
   - Observe that there are two ranges.
2. Remove the index.
   - Observe that the schema change job starts waiting for MVCC GC.
   - Observe that there are still 2 ranges.
   - Observe (via some ad-hoc logging) that new type of `GCHint` is written for the index range.
3. Restart the cluster with a `vBefore` binary.
4. Reset the sticky bit, and trigger merge of the two ranges.
   - Observe that the merge happened.
   - Observe that the `GCHint` was dropped/lost during this merge because the old code does not know about the new fields.
   - Observe that MVCC GC job still waits, and probably is stuck.
5. Restart the cluster with a `vAfter` binary.
   - Observe that eventually the schema change job does another MVCC range tombstone write to the index range (I was surprised to see this, in a good way).
   - Observe that new `GCHint` is written.
   - Observe that eventually MVCC GC removes the data, and the schema change job succeeds.

Given this experiment, I suspect that there is an easier mitigation for stuck jobs which wrote `vBefore` hints. Best case: when the cluster is restarted on `vAfter` binaries, the job is restarted last - then all the ranges it talks to will write new hints, and we're good. Possible case: the job is restarted and casts deletions again, but the ranges are still under `vBefore`. Then we will be stuck.
Mitigation: restart the job. Is that possible via console or CLI?

erikgrinaker commented:
>Mitigation: restart the job. Is that possible via console or CLI?

Yes, it can be paused and resumed.

pavelkalinnikov commented:
> Another case we should test is sending a snapshot from a new node to an old node, and vice versa. Can be done with `ALTER RANGE ... RELOCATE`. Also, running consistency checks on the range.

Done. Tested sending a snapshot both ways, and running consistency check afterwards.

> Yes, it can be paused and resumed: `PAUSE JOB x; RESUME JOB x;`

Also checked that resuming a job rewrites GC hints, and fixes a stalled schema change job.

Some details of the test:

```bash
roachprod create -n4 local
roachprod stage local:1-3 release v23.1.9
roachprod put local:4 cockroach # master + this PR

cockroach sql --insecure -f create.sql --url=$PGURL  # create table+index
cockroach sql --insecure -f show.sql --url=$PGURL    # see where leaseholder is
# Make sure leaseholder is not on node 4. Move it otherwise.
cockroach sql --insecure -f dropidx.sql --url=$PGURL # drop index

cockroach debug send-kv-batch req.json --insecure --url=$PGURL # remove sticky bit
# Then trigger merge.
# Then write many keys to the table so that GC score is low.

ALTER RANGE 65 RELOCATE VOTERS FROM 1 TO 4;
ALTER RANGE 65 RELOCATE LEASE TO 4;

SELECT * FROM crdb_internal.check_consistency(true, '', '') WHERE status != 'RANGE_CONSISTENT';                                                                                                
  range_id | start_key | start_key_pretty | status | detail | duration
-----------+-----------+------------------+--------+--------+-----------
(0 rows)

PAUSE JOB <id>;
RESUME JOB <id>;

# The job will write range tombstones again, now via node 4 which
# is on the new version. This will write new GC hints. Verified this
# with ad-hoc logging, seeing a GCHint with a non-zero timestamp in
# the new field.

# Now we're sure GC will respect this hint, so we wait 10 minutes.

# I also moved the voters and leaseholder for this range back to
# node 1, and ran consistency checks successfully again, to make
# sure that the old nodes process a snapshot with the new hint ok.

# Moved the voters and leases back to node 4. The index was GC-ed,
# and the schema change job completed.


```

erikgrinaker commented:
Cool, appreciate the additional testing!

pavelkalinnikov commented:
bors r=erikgrinaker
